### PR TITLE
Fix crashes related to handling of constexpr cast involving vector types

### LINKED ISSUE #2445: Implicit truncation of vector to constant causes crash
Trying to compile the following invalid shader will cause a crash inside of `ConstExprEmitter::VisitCastExpr` instead of a compile warning/error.

```hlsl
float4 main() : SV_Target0
{
	const float error = float4(0,0,0,0);
	return error;
}
```

### LINKED ISSUE #3522: const expr cast involving vector type fails in various scenarios
**Scenario 1: Const expr cast between vectors of same component count, but different component types causes compiler crash.** 

Below is the shader and  cmdlines which currently crashes.

%dxc -E main -T vs_6_2 -DVEC1 -DLTV=uint16_t1 -DRTV=uint1
%dxc -E main -T vs_6_2 -enable-16bit-types -DVEC1 -DLTV=uint1 -DRTV=uint16_t1
%dxc -E main -T vs_6_2 -DVEC3 -DLTV=bool3 -DRTV=int3

```
RWByteAddressBuffer rwbab;

void main() : OUT {

#ifdef VEC1
 const LTV v = RTV(0);
 rwbab.Store<LTV>(256, v);
 
#elif VEC2 
 const LTV v = RTV(1, 2);
 rwbab.Store<LTV>(512, v);
 
#elif VEC3
 const LTV v = RTV(4, 5, 6);
 rwbab.Store<LTV>(1024, v);
 
#else
 const LTV v = RTV(7, 8, 9, 10);
 rwbab.Store<LTV>(2048, v);
#endif
}
```

**Scenario 2: Const expr cast between vector-vector and vector-scalar causes compiler crash.** 

Below is an example shader and the cmdlines

%dxc -E main -T vs_6_2 -DLTV=uint -DRTV=uint4
%dxc -E main -T vs_6_2 -enable-16bit-types -DLTV=float16_t -DRTV=float16_t4
%dxc -E main -T vs_6_2 -enable-16bit-types -DLTV=bool3 -DRTV=float16_t4

```
RWByteAddressBuffer rwbab;

void main() : OUT {

 // Case 1: all zero constant
 const LTV v1 = RTV(0, 0, 0, 0);
 rwbab.Store<LTV>(100, v1);
 
 // Case 2: Non-zero constant
 const LTV v2 = RTV(1, 2, 3, 4);
 rwbab.Store<LTV>(200, v2);

}
```

### PR DESCRIPTION
**Pull request by vcsharma**
This PR fixes three different categories of bugs related to constexpr cast involving vectors.  

- Handle component cast of the following kinds:

`CK_HLSLCC_IntegralCast`
`CK_HLSLCC_IntegralToBoolean`
`CK_HLSLCC_FloatingToBoolean`

- Handle Vector truncation cast of the kind: `CK_HLSLVectorTruncationCast`

- Handle a special case of vector truncation where all elements of a vector are zero.

Fixes #2445
Fixes #3522

### CODE REVIEW

File: tools/clang/test/HLSLFileCheck/hlsl/types/cast/constexpr/constexpr_cast_vec_types_same_comp_count.hlsl [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ rwbab.Store<LTV>(512, v);
+ 
+#elif VEC3
+ const LTV v = RTV(4, 5, 6);
+ rwbab.Store<LTV>(1024, v);
+ 
+#else
+ const LTV v = RTV(7, 8, 9, 10);
+ rwbab.Store<LTV>(2048, v);
+#endif
```
Comment (pow2clk) [inline tools/clang/test/HLSLFileCheck/hlsl/types/cast/constexpr/constexpr_cast_vec_types_same_comp_count.hlsl:50]: An advantage of the longer vector declaration as I suggested above is that these ifdefs won't be needed.

Alternately, you could use the `##` preprocessor operator to paste TYPE and NUM tokens together

**Review by pow2clk — COMMENTED** [review]
A number of nitpicky alternative suggestions you can take or leave.

The functionality looks fine though it's obvious it doesn't play nice somehow with the SPIRV backend. Likely conflicting with this #502. @ehsannas and @jaebaek might have some suggestions

**Comment by jaebaek** [comment]
I will take a look and get back to you.

**Comment by vcsharma** [comment]
> I will take a look and get back to you.

Thanks @jaebaek!

**Comment by jaebaek** [comment]
Just sent a commit. Please let me know if there are other issues!

File: tools/clang/lib/AST/ExprConstant.cpp [hunk 2]
```diff
@@ -5732,6 +5732,60 @@ bool VectorExprEvaluator::VisitCastExpr(const CastExpr* E) {
...
+    return Success(Elts, E);
+  }
+  case CK_HLSLCC_FloatingToBoolean:
+  case CK_HLSLCC_IntegralToBoolean: {
+    if (!Visit(SE))
+      return Error(E);
+    SmallVector<APValue, 4> Elts;
+    for (uint32_t i = 0; i < Result.getVectorLength(); ++i) {
+      APValue Elem = Result.getVectorElt(i);
+      bool Result;
```
[the remark this answers is not in the document]
Comment (vcsharma) [reply tools/clang/lib/AST/ExprConstant.cpp]: Good catch, fixed.

File: tools/clang/test/HLSLFileCheck/hlsl/types/cast/constexpr/constexpr_cast_vec_to_scalar_type.hlsl [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// RUN: %dxc -E main -T vs_6_2 -enable-16bit-types -DLTV=uint -DRTV=uint4 %s | FileCheck %s
```
[the remark this answers is not in the document]
Comment (vcsharma) [reply tools/clang/test/HLSLFileCheck/hlsl/types/cast/constexpr/constexpr_cast_vec_to_scalar_type.hlsl]: Renamed the test.

**Comment by vcsharma** [comment]
> Just sent a commit. Please let me know if there are other issues!

Thanks for fixing the SPIRV failure!

**Pull request merged**